traversal/producer/GraphProducer.java

```diff
@@ -46,91 +47,102 @@
+            ResourceIterator<VertexMap> iter =
+                    new GraphIterator(graphMgr, start.next(), procedure, params, filter).distinct(produced);
+            runningJobs.put(iter, CompletableFuture.completedFuture(null));
         }
-        if (iteratorRequested.size() == 0) {
-            done(queue);
-        } else {
-            int splitCount = (int) Math.ceil((double) count / iteratorRequested.size());
-            for (ResourceIterator<VertexMap> iterator : iteratorRequested.keySet()) {
-                assert iteratorRequested.containsKey(iterator);
-                iteratorRequested.computeIfPresent(iterator, (k, v) -> v + splitCount);
-                if (!iteratorJobs.containsKey(iterator)) {
-                    iteratorJobs.put(iterator, runAsync(() -> produceAsync(queue, iterator), forkJoinPool()));
-                }
-            }
+        isInitialised = true;
+        if (runningJobs.isEmpty()) done(queue);
+    }
+
+    private synchronized void distribute(Queue<VertexMap> queue, int request) {
+        if (isDone.get()) return;
+        int requestSplitMax = (int) Math.ceil((double) request / runningJobs.size());
+        int requestSent = 0;
+        for (ResourceIterator<VertexMap> iterator : runningJobs.keySet()) {
+            int requestSplit = Math.min(requestSplitMax, request - requestSent);
+            runningJobs.computeIfPresent(iterator, (iter, asyncJob) -> asyncJob.thenRunAsync(
+                    () -> job(queue, iter, requestSplit), forkJoinPool()
+            ));
+            requestSent += requestSplit;
+            if (requestSent == request) break;
         }
     }
 
-    private synchronized int take(ResourceIterator<VertexMap> iterator) {
-        int count = iteratorRequested.get(iterator);
-        iteratorRequested.put(iterator, 0);
-        return count;
+    private synchronized void transition(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator, int unfulfilled) {
+        if (iterator.hasNext()) return;
+        if (runningJobs.containsKey(iterator) && start.hasNext()) replace(queue, iterator, unfulfilled);
+        else if (!runningJobs.isEmpty() && unfulfilled > 0) distribute(queue, unfulfilled);
+        else if (runningJobs.isEmpty()) done(queue);
     }
 
-    private synchronized void compensate(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator, int requested) {
-        iteratorJobs.remove(iterator);
-        int toCompensate = requested + take(iterator);
-        if (!iterator.hasNext()) iteratorRequested.remove(iterator);
-        if (toCompensate > 0) produce(queue, toCompensate);
+    private synchronized void replace(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator, int unfulfilled) {
+        runningJobs.remove(iterator);
+        ResourceIterator<VertexMap> newIter =
+                new GraphIterator(graphMgr, start.next(), procedure, params, filter).distinct(produced);
```

Comment:
Maybe assert `start.hasNext()` here. It's not clear what the preconditions of this function are.

Reply:
I've removed `iterator` from the input completely now